include/envoy/http/access_log.h

```diff
@@ -192,6 +195,41 @@ class Formatter {
 
 typedef std::unique_ptr<Formatter> FormatterPtr;
 
+/**
+ * Implemented for each AccessLog::Instance and registered via Registry::registerFactory or the
+ * convenience class RegisterFactory.
+ */
+class AccessLogInstanceFactory {
+public:
+  virtual ~AccessLogInstanceFactory() {}
+
+  /**
+   * Create a particular AccessLog::Instance implementation from a config proto. If the
+   * implementation is unable to produce a factory with the provided parameters, it should throw an
+   * EnvoyException. The returned pointer should never be nullptr.
+   * @param config the custom configuration for this access log type.
+   * @param filter filter to determine whether a particular request should be logged. If no filter
+   * was specified in the configuration, argument will be nullptr.
+   * @param log_manager interface through which filesystem access logs are created.
```

Comment:
Is this the only way to plumb this for the filesystem logs? Seems like a vestigial feature for everyone else.

Reply:
We could do something more general, like provide an "environment" or "context" to give general access to conditionally useful variables so that this api can be expanded with new use cases (for instance, we will likely need to provide the ability to access the cluster manager for RPC logs) as is done for http filters now.  We could also make the log_manager a singleton, but that seems like it would be more troublesome for each additional object.

Reply:
+1 I think a context is good, it leaves room for future expansion without having to modify the filter interface, which causes breaking/deprecating changes.